README.md
# NEM Wallet #

### How to use the NEM Wallet ? ###

After downloading the latest version here https://forum.nem.io/t/nano-wallet-beta-1-1-11-5-000-xem-bug-bounty/2791, extract the NEM-Wallet folder from the archive, then you can simply click on start.html to open the app in your default browser. No installation or local server needed.

### How to import NCC wallet ? ###

Open NCC and your wallet, clicking on user icon should reveal a drop-down menu with "Export for Lightwallet".
Once your have your wallet (.json format), open the NEM Wallet and import the file from the login page. It'll ask for an upgrade and trigger the download of your updated wallet.
Do not delete your old NCC wallet and keep it safe somewhere.

### Imported accounts that are already harvesting using NCC ###

NCC is using a different way to generate the remote associated with an account. NEM Wallet uses BIP32 to generate it deterministically and make everything easier. In order to manage and monitor harvesting, you'll need to deactivate the old remote account from NCC and activate your new remote account from the NEM Wallet. Alternatively you can use a custom delegated public key in the NEM Wallet importance transaction page.

**/!\ Main network: Please note that if you are part of the Supernodes program you'll need to re-enroll your node using your new delegated public key.**

### How to update the NEM Wallet ? ###

Before update make sure that you don't have any wallets stored into the NEM-Wallet folder (you can place them anywhere you want). Then, just delete old version and unzip new version.

###  Dashboard is empty and no balance is showing ###

It is probably that the node you are connected to is down.
Look at the navigation bar, at the top of the page.
If the circle next to `Node` is red, click on it and select another node from the list.

# Developers #

### Build from source ###

1) Install gulp

<pre>npm install -g gulp-cli</pre>

2) Open a console to the path of the NEM-Wallet folder and install all the needed dependencies

<pre>npm install</pre>

3) Build:

<pre>gulp</pre>

##### OR #####

4) Run the NEM Wallet as App (packaged with chromium browser)

<pre>npm run devApp</pre>

##### OR #####

5) Build NEM Wallet apps, default set to : win64,osx64,linux64

<pre>gulp build-app</pre>

6) For local use, build NEM Wallet Electron apps (only Electron verision support Ledger wallets), default build for MacOS, Windows and Linux

<pre>
# create the release folder where the artifacts will be created
mkdir -p release

# to skip code signing
export CSC_IDENTITY_AUTO_DISCOVERY=false

npm run release
</pre>

7) Release for distribution: (Code signing for Apple builds - requires `Developer ID Certificate`)

    7.1 On a MacOS machine, download the zip file containing the app signing certificates (ask team)

    7.2 Extract the certificates and double click each one of them to add to the keychain (ask the team for private key password)
    
    7.3 If `Developer ID Certificate` is signing the app for the first time then the app needs to be notarized and this is an automated step in the process.You'll need to enable notarization by setting the following env vars.(skip this step already signed before)

    <pre>
    export DESKTOP_APP_NOTARIZE=true
    export DESKTOP_APP_APPLE_ID=VALID_APPLE_DEV_ID
    export DESKTOP_APP_APPLE_PASSWORD=VALID_APPLE_DEV_PASSWORD
    </pre>
    
    7.4 Enable auto discovery for code signing process to pick up the certificates from the keychain

    <pre>export CSC_IDENTITY_AUTO_DISCOVERY=true</pre>

    7.5 Run release
    <pre>npm run release</pre>

    7.6 Validate if the app is signed with a Developer ID Certificate
    
    <pre>spctl -a -t exec -v ./release/mac/Nem\ Wallet.app</pre>
        

### Known issues ###

- Sometimes, depending the node used, unconfirmed data to sign is not incoming from websocket


CHANGELOG.md
# Changelog

## Version 2.7.0

    - Added SuperNode program module
        - It allows user to enroll the SuperNode program from the wallet
        - It allows the user to track reward payout history
        - It shows node harvesting status and total collected rewards
    - Fixed app code signing and notarization issues and updated the Readme

## Version 2.6.2

    - Upgraded trezor-connect to v8
    - Refactored voting module
    - Added unit tests for Trezor
    - Removed dependency to nem-trezor lib
    - Fixed failing unit tests
    - Fixed voting transactions fee flooring issue

## Version 2.6.1

    - Added post-launch opt-in check NIS keys

## Version 2.6.0

    - Added post-launch opt-in process

## Version 2.5.1

    - Fixed Ledger opt-in cosigning issue.

## Version 2.5.0

    - Added Electron builds
    - Added Ledger support
    - Fixed optin minRemoval calculation

## Version 2.4.8

    - Fix voting service
    - Updated company details in terms and condition

## Version 2.4.7-OptIn

    - Added Symbol opt in module

## Version 2.4.7

    - Fix testnet nodes

## Version 2.4.6

    - Fix Changelly widget
    - Fix voting
    - Fix broken node link in faq

## Version 2.4.4

    - Updated voting module
    - Fix default testnet node

## Version 2.4.2

    - Updated Voting module
    - Added NEMonster collection game (by @aenima86)
    - Fixed Changelly widget
    - Renamed to NEM Wallet
    - Minor fixes and improvements

## Version 2.3.2

    - Fix DNS search
    - Update languages
    - Minor improvements

## Version 2.3.0

    - Domain Name System module
    - Import Wallet QR code module
    - Support voting with Trezor
    - Minor fixes and improvements

## Version 2.2.0

    - Create and verify signed messages modules
    - Account transactions explorer
    - Fix timestamp too far in the future
    - Updated voting backend
    - Minor fixes and improvements

## Version 2.1.2

    - Voting modules
    - Offline transaction modules
    - Advanced Apostille
    - BIP32 refactor
    - News and transaction confirmation components (WIP)
    - Invoice module and component
    - Minor fixes and improvements

## Version 2.0.14

    - Fix brain wallet login

## Version 2.0.13

    - Hide Trezor login button in Chrome builds

## Version 2.0.12

    - Fix harvesting on mainnet
    - Fix new update notification

## Version 2.0.10

    - Fix Trezor support
    - Minor fixes and improvements

## Version 2.0.2

    - Fix password input showing if using Trezor
    - Shows encryption fee message in fee input

## Version 2.0.0

    - Use the [nem-sdk](https://github.com/QuantumMechanics/NEM-sdk)
    - Trezor support
    - ShapeShift support
    - Step by step signup and account safety protocol
    - Extra entropy added to simple wallets
    - Added FAQ
    - Can show balance and market data in different currencies
    - Notification on homepage when new update available

## Version 1.4.13

    - Allow sub-namespaces and mosaic names to start with a number
    - Fix secondary accounts: In **occasional** cases, secondary accounts private keys can start with '00' but BIP32 was omitting it, giving a 62 characters private key. Those **secondary accounts** will fail to send anything out even if password is valid (but recoverable).
    - Update languages

## Version 1.4.10

    - New fee structure on Mainnet
    - Added passphrase strength estimator to brain wallet: https://github.com/dropbox/zxcvbn
    - Brain wallet passphrase requires 40 characters minimum
    - Moved purge button into footer
    - Show an alert asking to change node if not connected within 5 seconds after login
    - Show an alert if user's brain wallet seems weak (< 40 characters)
    - Minor fixes and improvements

## Version 1.4.3

    - Voting module on Mainnet
    - Removed bob.nem.ninja and used MedAlice2 as default testnet node
    - Minor fixes and improvements

## Version 1.4.0

    - New fee structure (on testnet)
    - Voting module (on testnet), made by @shierve https://github.com/shierve
    - If user deny geoloc for nearest node, app will select a random supernode
    - Fixed languages
    - Minor fixes and improvements

## Version 1.3.12

    - Improve start / stop delegated harvesting design
    - Fix market data percentage change
    - Allow only one cosignatory removal per multisig edit
    - Prevent multisig edit to make a 0 of m (= m of m)
    - Module to sign multisig transactions if not pushed into your unconfirmed transactions
    - Fetch the closest node from user location (https://www.w3schools.com/html/html5_geolocation.asp26)
    - Minor fixes and improvements

## Version 1.3.4

    - New account safety measures at wallet creation
    - Transaction details and multisignature dropdowns shows account label if present in address book
    - Selected language stored in local storage
    - Russian translation
    - Minor fixes & improvements

## Version 1.3.0

    - New design
    - Address book
    - Fix Apostille on Chrome and Safari
    - No need to provide an account address to create a private key wallet, address will be shown automatically
    - Same as above point but for creating multisig accounts
    - Importance transaction module show the corresponding remote account address when a custom public key is provided
    - Fix handling of decimal amount in normal transfer transaction, comma and dot decimal mark can be used
    - Fix message fee
    - Fix 'unknown mosaic divisibility' error when receiving a new mosaic
    - Limit individual apostille file to 100MB
    - Fix display of importance score
    - Fix "FAILURE_TIMESTAMP_TOO_FAR_IN_FUTURE" using NIS time-sync API
    - Lock app and show message if browser not supported
    - New market data provider
    - Minor fixes & improvements

## Version 1.2.12

    - Renew namespace module
    - Renewal alert when namespaces can be renewed (one month before expiry)
    - Fix multisignature module

tests/specs/trezor.spec.js
import { fail } from 'assert';
import { TrezorAccount } from '../../src/app/modules/trezor/trezorAccount';

describe('Trezor integration', function () {
    const transaction = {
        type: 257,
        version: -1744830463,
        timeStamp: 130922553,
        deadline: 130926153,
        recipient: 'TA545ICAVNEUDFUBIHO3CEJBSVIZ7YYHFFX5LQPT',
        amount: 1000000,
        fee: 100000,
        message: {
            type: 1,
            payload: '616263',
        },
        mosaics: null,
        setNetworkType: function (network) {
            this.network = network;
        },
        toDTO: function () {},
    };
    const address = 'TA545ICAVNEUDFUBIHO3CEJBSVIZ7YYHFFX5LQPT';
    const hdKeyPath = "44'/43'/152'/0'/0'";
    const network = -104;
    const Trezor = {
        serialize: () => {},
    };

    it('Can serialize single transaction for trezor account', async function (done) {
        // Arrange
        const trezorAccount = new TrezorAccount(address, hdKeyPath, network, Trezor);
        let expectedResult = 'payload';
        spyOn(Trezor, 'serialize').and.returnValue(Promise.resolve(expectedResult));

        let result;

        // Act
        try {
            result = await trezorAccount.signTransaction(transaction).toPromise();
        } catch (err) {
            fail('Error while serializing single transaction for trezor: ' + err);
        }

        // Assert
        expect(Trezor.serialize).toHaveBeenCalled();
        expect(result).toEqual(expectedResult);
        done();
    });

    it('Can serialize multiple transactions for trezor account', async function (done) {
        // Arrange
        const trezorAccount = new TrezorAccount(address, hdKeyPath, network, Trezor);
        let expectedResults = ['payload 1', 'payload 2'];
        spyOn(Trezor, 'serialize').and.returnValues(...expectedResults.map((r) => Promise.resolve(r)));

        let result;

        // Act
        try {
            result = await trezorAccount.signTransactions([transaction, transaction]).first().toPromise();
        } catch (err) {
            fail('Error while serializing multiple transactions for trezor: ' + err);
        }

        // Assert
        expect(Trezor.serialize).toHaveBeenCalled();
        expect(result).toEqual(expectedResults);
        done();
    });
});


scripts/notarize.js
const { notarize } = require('electron-notarize');
var appBundleId = require('../package.json').build.appId;


// You will need to notarize the application if the "Developer ID Certificate" is first time being used
exports.default = async function notarizing(context) {
  const { electronPlatformName, appOutDir } = context;  
  if (electronPlatformName !== 'darwin') {
    console.log('Skipping notarizing because this is not a macOS build.');
    return;
  } else if (process.env.DESKTOP_APP_NOTARIZE !== 'true') {
    console.log('Skipping notarizing because DESKTOP_APP_NOTARIZE env is not set.');
    return;
  } else if (process.env.DESKTOP_APP_APPLE_ID === undefined || process.env.DESKTOP_APP_APPLE_PASSWORD === undefined) {
    console.log('Skipping notarizing because DESKTOP_APP_APPLE_ID or DESKTOP_APP_APPLE_PASSWORD env are not set.');
    return;
  }
  const appName = context.packager.appInfo.productFilename;
  const appPath = `${appOutDir}/${appName}.app`;

  console.log(`Notarizing ${appName} with bundleId[${appBundleId}] at ${appPath}`);

  return await notarize({
    appBundleId: appBundleId,
    appPath: appPath,
    appleId: process.env.DESKTOP_APP_APPLE_ID,
    appleIdPassword: process.env.DESKTOP_APP_APPLE_PASSWORD,
  });
};


src/app/services/voting.service.js
import {Observable} from 'rxjs';
import {TrezorAccount} from '../modules/trezor/trezorAccount';
import nemsdk from 'nem-sdk';
const nem = require('nem-library');
const voting = require('nem-voting');

class Voting {
    constructor($filter, $timeout, Alert, Wallet, Ledger, Trezor, VotingUtils, DataStore) {
        'ngInject';

        /***
         * Declare services
         */
        this._$timeout = $timeout;
        this._$filter = $filter;
        this._Alert = Alert;
        this._Wallet = Wallet;
        this._Ledger = Ledger;
        this._Trezor = Trezor;
        this._DataStore = DataStore;
        this._VotingUtils = VotingUtils;
        if(this._Wallet.network < 0){
            nem.NEMLibrary.bootstrap(nem.NetworkTypes.TEST_NET);
        }
        else{
            nem.NEMLibrary.bootstrap(nem.NetworkTypes.MAIN_NET);
        }
    }

    // Voting Functions

    init() {
        if ((this._Wallet.network < 0 && nem.NEMLibrary.getNetworkType() !== nem.NetworkTypes.TEST_NET) ||
            (this._Wallet.network > 0 && nem.NEMLibrary.getNetworkType() !== nem.NetworkTypes.MAIN_NET)) {
                nem.NEMLibrary.reset();
                if(this._Wallet.network < 0){
                    nem.NEMLibrary.bootstrap(nem.NetworkTypes.TEST_NET);
                }
                else{
                    nem.NEMLibrary.bootstrap(nem.NetworkTypes.MAIN_NET);
                }
            }
    }

    /**
     * getPolls(pollIndexAddress) returns a list with the poll headers from all the polls that are on the given index
     *
     * @param {string} pollIndexAddress - NEM address for the poll index account
     * @param {string} lastId - the id of the last transaction fetched for pagination
     *
     * @return {promise} - a list of all the poll header objects on the index account
     */
    getPolls(pollIndexAddress, lastId) {
        this.init();
        const obs = voting.PollIndex.fromAddress(new nem.Address(pollIndexAddress), lastId)
            .map((index) => {
                const headers = index.headers.map((header) => {
                    return {
                        title: header.title,
                        type: header.type,
                        doe: header.doe,
                        address: header.address.plain(),
                        creator: header.creator.plain(),
                        whitelist: header.whitelist,
                    }
                });
                return {
                    polls: headers,
                    lastId: index.lastId,
                }
            });
        return obs.first().toPromise();
    }

    /**
     * getOfficialPolls(pollIndexAddress) returns a list with the poll headers for all the official polls
     *
     * @return {promise} - a list of all the poll header objects on the index account
     */
    getOfficialPolls() {
        this.init();
        const obs = voting.getAllOfficialPolls()
            .map((polls) => {
                return polls.map((header) => {
                    return {
                        title: header.title,
                        type: header.type,
                        doe: header.doe,
                        address: header.address.plain(),
                        creator: header.creator.plain(),
                        whitelist: header.whitelist,
                    }
                });
            });
        return obs.first().toPromise();
    }

    /**
     * createPoll(details, pollIndex, common) creates a poll with the given details on the given pollIndex
     *
     * @param {object} details - poll details, without the option addresses
     * @param {string} pollIndex - NEM address of the poll index
     * @param {object} common - common object
     *
     * @return {promise} - the generated poll Address
     */
    createPoll(details, pollIndex, common) {
        this.init();
        let d = new Date();
        let chainTime = this._DataStore.chain.time;
        console.log("chain time:", chainTime);
        let timeStamp = Math.floor(chainTime) + Math.floor(d.getSeconds() / 10);
        let due = (this._Wallet.network === nemsdk.model.network.data.testnet.id) ? 60 : 24 * 60;
        let deadline = timeStamp + due * 60;
        const formData = {
            title: details.formData.title,
            doe: details.formData.doe,
            type: details.formData.type,
            multiple: details.formData.multiple,
        };
        const description = details.description;
        const options = details.options;
        const whitelist = (details.formData.type === 1) ? details.whitelist.map(a => new nem.Address(a)) : undefined;

        const poll = new voting.UnbroadcastedPoll(formData, description, options, whitelist);

        let account = {};
        if (common.isHW) {
            if (this._Wallet.algo == "trezor") {
                account = new TrezorAccount(this._Wallet.currentAccount.address, this._Wallet.currentAccount.hdKeypath, this._Trezor);
            }
        } else {
            account = nem.Account.createWithPrivateKey(common.privateKey);
        }

        const index = new voting.PollIndex(new nem.Address(pollIndex), false, []);

        const broadcastData = poll.broadcast(account.publicKey, index);
        broadcastData.transactions = broadcastData.transactions.map((t) => {
            t.timeWindow = nem.TimeWindow.createFromDTOInfo(timeStamp, deadline);
            t.fee = Math.floor(t.fee);
            return t;
        });
        const nodeSplit = this._Wallet.node.host.split("://");
        const node = {
            protocol: nodeSplit[0],
            domain: nodeSplit[1],
            port: this._Wallet.node.port,
        }
        console.log("node", node);
        const transactionHttp = new nem.TransactionHttp([
            node,
        ]);
        const signTransaction = (i) => {
            let p;
            if (common.isHW) {
                if (this._Wallet.algo == "trezor") {
                    if (window['isElectronEnvironment']) {
                        const transaction = broadcastData.transactions[i];
                        transaction.setNetworkType(nem.NEMLibrary.getNetworkType());
                        const dto = transaction.toDTO();
                        p = this._Trezor.serialize(dto, this._Wallet.currentAccount);
                    } else {
                        p = account.signTransaction(broadcastData.transactions[i]).first().toPromise();
                    }
                } else if (this._Wallet.algo == "ledger") {
                    const transaction = broadcastData.transactions[i];
                    transaction.setNetworkType(nem.NEMLibrary.getNetworkType());
                    const dto = transaction.toDTO();
                    p = this._Ledger.serialize(dto, this._Wallet.currentAccount);
                }
            } else {
                p = Promise.resolve(account.signTransaction(broadcastData.transactions[i]));
            }
            return p.then((signed) => {
                if (broadcastData.transactions.length - 1 === i) {
                    return [signed];
                }
                if (common.isHW && this._Wallet.algo == "trezor" && window['isElectronEnvironment']) {
                    return new Promise(resolve => setTimeout(() => signTransaction(i + 1).then((next) => {
                        resolve([signed].concat(next));
                    }).catch(err => resolve([null])), 500));
                } else {
                    return signTransaction(i + 1).then((next) => {
                        return [signed].concat(next);
                    });
                }
            }).catch(err => {
                if (common.isHW && this._Wallet.algo == "ledger") {
                    throw 'handledLedgerErrorSignal';
                } else {
                    throw err;
                }
            });
        }

        return signTransaction(0).then((signedTransactions) => {
            return Observable.merge(...(signedTransactions.map((t) => {
                return transactionHttp.announceTransaction(t);
            })))
                .map(() => {
                    return broadcastData.broadcastedPoll.address.plain();
                }).last().toPromise();
        }).catch(err => {
            throw err;
        });

    }

    /**
     * vote(address, common, multisigAccount) sends a vote to the given address. Can vote as multisig
     *
     * @param {object} poll - poll address
     * @param {string} option - option string on which to vote
     * @param {object} common - common object
     * @param {string} multisigAccount - NEM address of the multisig account we want to send the vote for (optional)
     *
     * @return {promise} - returns a promise that resolves when the vote has been sent
     */
    vote(poll, option, common, multisigAccount) {
        this.init();
        let d = new Date();
        let chainTime = this._DataStore.chain.time;
        let timeStamp = Math.floor(chainTime) + Math.floor(d.getSeconds() / 10);
        let due = (this._Wallet.network === nemsdk.model.network.data.testnet.id) ? 60 : 24 * 60;
        let deadline = timeStamp + due * 60;
        return voting.BroadcastedPoll.fromAddress(new nem.Address(poll))
            .map((poll) => {
                let voteTransaction;
                if (multisigAccount) {
                    const multisigAcc = nem.PublicAccount.createWithPublicKey(multisigAccount.publicKey);
                    voteTransaction = poll.voteMultisig(multisigAcc, option);
                    voteTransaction.timeWindow = nem.TimeWindow.createFromDTOInfo(timeStamp, deadline);
                    voteTransaction.otherTransaction.timeWindow = nem.TimeWindow.createFromDTOInfo(timeStamp, deadline);
                } else {
                    voteTransaction = poll.vote(option);
                    voteTransaction.timeWindow = nem.TimeWindow.createFromDTOInfo(timeStamp, deadline);
                }
                return voteTransaction;
            }).first().toPromise();
    }

    /**
     * pollDetails(pollAddress) returns the details of a poll stored in the given pollAddress
     *
     * @param {string} pollAddress - NEM address for the poll account
     *
     * @return {promise} - a promise that returns the details object of the poll
     */
    pollDetails(pollAddress) {
        this.init();
        return voting.BroadcastedPoll.fromAddress(new nem.Address(pollAddress))
            .map((poll) => {
                const data = {
                    formData: poll.data.formData,
                    description: poll.data.description,
                    options: {
                        strings: poll.data.options,
                        link: {},
                    },
                    whitelist: poll.data.whitelist,
                };
                poll.data.options.forEach((option) => {
                    data.options.link[option] = poll.getOptionAddress(option).plain();
                });
                return data;
            }).first().toPromise();
    }

    /**
     * getResults(pollAddress, type, end) returns the result object for the poll depending of the type of the counting
     *
     * @param {string} pollAddress - NEM address of the poll
     *
     * @return {promise} - A promise that returns the result object of the poll
     */
    getResults(pollAddress) {
        this.init();
        return voting.BroadcastedPoll.fromAddress(new nem.Address(pollAddress))
            .switchMap((poll) => {
                return poll.getResults();
            }).first().toPromise();
    }

    /**
     * hasVoted(addrses, pollDetails)
     *
     * @param {string} address - NEM address of the poll
     * @param {object} pollDetails - poll details object of the poll. The details can be obtained from the address,
     * but passing as a parameter is faster, since when we check for votes on the voting module we already have the details
     *
     * @return {promise} - A promise that returns:
     *                          0 if there are no votes
     *                          1 if there is an unconfirmed vote
     *                          2 if there is a confirmed vote
     */
    hasVoted(address, pollDetails) {
        this.init();
        var orderedAddresses = [];
        if(pollDetails.options.link){
            orderedAddresses = pollDetails.options.strings.map((option)=>{
                return pollDetails.options.link[option];
            });
        }
        else{
            orderedAddresses = pollDetails.options.addresses;
        }
        var confirmedPromises = orderedAddresses.map((optionAddress) => {
            return this._VotingUtils.existsTransaction(address, optionAddress);
        });
        return Promise.all(confirmedPromises).then((data) => {
            return Math.max.apply(null, data);
        });
    }

    isInWhitelist(address, whitelist) {
        address = new nem.Address(address);
        const findI = whitelist.findIndex(a => a.plain() === address.plain());
        return findI >= 0;
    }

    getBroadcastFee(details) {
        this.init();
        const formData = {
            title: details.formData.title,
            doe: details.formData.doe,
            type: details.formData.type,
            multiple: details.formData.multiple,
        };
        const description = details.description;
        const options = details.options;
        let whitelist;
        if (details.formData.type === 1) {
            console.log("wl", details.whitelist);
            try {
                whitelist = details.whitelist.map(a => new nem.Address(a));
            } catch (err) {
                return 0;
            }
        }

        const poll = new voting.UnbroadcastedPoll(formData, description, options, whitelist);

        return poll.getBroadcastFee();
    }

    broadcastVotes(votes, common) {
        let account;
        if (common.isHW) {
            if (this._Wallet.algo == "trezor") {
                account = new TrezorAccount(this._Wallet.currentAccount.address, this._Wallet.currentAccount.hdKeypath, this._Trezor);
            }
        } else {
            account = nem.Account.createWithPrivateKey(common.privateKey);
        }
        // sign
        let signedTransactionsPromise;
        if (common.isHW) {
            if (this._Wallet.algo == "trezor") {
                if (window['isElectronEnvironment']) {
                    const signTransaction = (i) => {
                        const transaction = votes[i];
                        transaction.setNetworkType(nem.NEMLibrary.getNetworkType());
                        const dto = transaction.toDTO();
                        const p = this._Trezor.serialize(dto, this._Wallet.currentAccount);
                        return p.then((signed) => {
                            if (votes.length - 1 === i) {
                                return [signed];
                            }
                            return new Promise(resolve => setTimeout(() => signTransaction(i + 1).then((next) => {
                                resolve([signed].concat(next));
                            }).catch(err => resolve([null])), 500));
                        }).catch(err => {
                            throw err;
                        });
                    }
                    signedTransactionsPromise = signTransaction(0);
                } else {
                    signedTransactionsPromise = account.signTransactions(votes).first().toPromise();
                }
            } else if (this._Wallet.algo == "ledger") {
                const signTransaction = (i) => {
                    const transaction = votes[i];
                    transaction.setNetworkType(nem.NEMLibrary.getNetworkType());
                    const dto = transaction.toDTO();
                    const p = this._Ledger.serialize(dto, this._Wallet.currentAccount);
                    return p.then((signed) => {
                        if (votes.length - 1 === i) {
                            return [signed];
                        }
                        return signTransaction(i + 1).then((next) => {
                            return [signed].concat(next);
                        });
                    }).catch(err => {
                        throw 'handledLedgerErrorSignal';
                    });
                }
                signedTransactionsPromise = signTransaction(0);
            }
        } else {
            const signed = votes.map(v => {
                return account.signTransaction(v);
            });
            signedTransactionsPromise = Promise.resolve(signed);
        }
        // broadcast
        return signedTransactionsPromise.then(signedTransactions => {
            const nodeSplit = this._Wallet.node.host.split("://");
            const node = {
                protocol: nodeSplit[0],
                domain: nodeSplit[1],
                port: this._Wallet.node.port,
            }
            const transactionHttp = new nem.TransactionHttp([
                node,
            ]);
            return Promise.all(signedTransactions.map(t => {
                return transactionHttp.announceTransaction(t).first().toPromise();
            }));
        }).catch(e => {
            throw e;
        });
    }
}

export default Voting;


src/app/modules/trezor/trezor.service.js
import { TransactionTypes } from 'nem-library';
import nem from 'nem-sdk';
import TrezorConnect from 'trezor-connect';

/** Service storing Trezor utility functions. */
class Trezor {

    /**
     * Initialize dependencies and properties
     *
     * @params {services} - Angular services to inject
     */
    constructor() {
        'ngInject';

        // Service dependencies region //

        // End dependencies region //

        // Service properties region //

        // End properties region //
        TrezorConnect.manifest({ email: 'dev@symbol.dev', appUrl: 'https://symbol.dev' });
    }

    // Service methods region //

    createWallet(network) {
        return this.createAccount(network, 0, "Primary").then((account) => ({
            "name": "TREZOR",
            "accounts": {
                "0": account
            }
        }));
    }

    bip44(network, index) {
        const coinType = network == -104 ? 1 : 43;

        return `m/44'/${coinType}'/${index}'/0'/0'`;
    }

    createAccount(network, index, label) {
        return new Promise((resolve, reject) => {
            const hdKeypath = this.bip44(network, index);
            TrezorConnect.nemGetAddress({
                path: hdKeypath,
                network: network < 0 ? 256 + network : network,
                showOnTrezor: true
            }).then(function(result) {
                if (result.success) {
                    resolve({
                        "brain": false,
                        "algo": "trezor",
                        "encrypted": "",
                        "iv": "",
                        "address": result.payload.address,
                        "label": label,
                        "network": network,
                        "child": "",
                        "hdKeypath": hdKeypath
                    });
                } else {
                    reject(result.payload.error);
                }
            })
        });
    }

    deriveRemote(account, network) {
        const key = "Export delegated harvesting key?";
        const value = "0000000000000000000000000000000000000000000000000000000000000000";

        return new Promise((resolve, reject) => {
            TrezorConnect.cipherKeyValue({
                path: account.hdKeypath,
                key: key,
                value: value,
                encrypt: true,
                askOnEncrypt: true,
                askOnDecrypt: true
            }).then(function(result) {
                if (result.success) {
                    const privateKey = nem.utils.helpers.fixPrivateKey(result.payload.value);
                    const keyPair = nem.crypto.keyPair.create(privateKey);
                    const publicKey = keyPair.publicKey.toString();
                    const address = nem.model.address.toAddress(publicKey, network);

                    resolve({
                        address,
                        privateKey,
                        publicKey
                    });
                } else {
                    reject(result.payload.error);
                }
            })
        });
    }

    /**
     * Adjusts importance transfer transaction to be compatible with Trezor
     * 
     * @param {Object} transaction - Transaction to adjust
     * @returns the adjusted transaction or the same transaction (if not an importance transfer)
     */
    adjustImportanceTransferTransaction(transaction) {
        if (transaction.type === TransactionTypes.IMPORTANCE_TRANSFER) {
          return {
              ...transaction,
              importanceTransfer: {
                mode: transaction.mode,
                publicKey: transaction.remoteAccount,
              },
            }
        }
        return transaction;
    }

    serialize(transaction, account) {
        const tx = this.adjustImportanceTransferTransaction(transaction);

        return new Promise((resolve, reject) => {
          TrezorConnect.nemSignTransaction({
            path: account.hdKeypath,
            transaction: tx,
          }).then(function (result) {
            if (result.success) {
              resolve(result.payload);
            } else {
              reject({
                code: 0,
                data: {
                  message: result.payload.error,
                },
              });
            }
          });
        });
    }

    showAccount(account) {
        return new Promise((resolve, reject) => {
            TrezorConnect.nemGetAddress({
                path: account.hdKeypath,
                network: account.network < 0 ? 256 + account.network : account.network,
                showOnTrezor: true
            }).then(function(result) {
                if (result.success) {
                    resolve(result.payload.address);
                } else {
                    reject(result.payload.error);
                }
            })
        });
    }

    // End methods region //

}

export default Trezor;
